File: python/src/scipp/_bins.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
# @author Simon Heybrock
from ._scipp import core as _cpp
from ._cpp_wrapper_util import call_func as _call_cpp_func


class lookup:
    def __init__(self, func: _cpp.DataArrayView, dim: str):
        self.func = func
        self.dim = dim


class _Bins:
    """
    Proxy for operations on bins of a variable
    """
    def __init__(self, obj):
        self._obj = obj

    def _data(self):
        try:
            return self._obj.data
        except AttributeError:
            return self._obj

    def __mul__(self, lut: lookup):
        copy = self._obj.copy()
        _cpp.buckets.scale(copy, lut.func, lut.dim)
        return copy

    def __truediv__(self, lut: lookup):
        copy = self._obj.copy()
        _cpp.buckets.scale(copy, _cpp.reciprocal(lut.func), lut.dim)
        return copy

    def __imul__(self, lut: lookup):
        _cpp.buckets.scale(self._obj, lut.func, lut.dim)
        return self

    def __itruediv__(self, lut: lookup):
        _cpp.buckets.scale(self._obj, _cpp.reciprocal(lut.func), lut.dim)
        return self

    @property
    def begin(self):
        """Begin index of bins as view of internal data buffer"""
        return _cpp.bins_begin_end(self._data())[0]

    @property
    def end(self):
        """End index of bins as view of internal data buffer"""
        return _cpp.bins_begin_end(self._data())[1]

    @property
    def dim(self):
        """Dimension of internal data buffer used for slicing into bins"""
        return _cpp.bins_dim(self._data())

    @property
    def data(self):
        """Internal data buffer holding data of all bins"""
        return _cpp.bins_data(self._data())

    def sum(self):
        """Sum of each bin.

        :return: The sum of each of the input bins.
        :seealso: :py:func:`scipp.sum` for summing non-bin data
        """
        return _call_cpp_func(_cpp.buckets.sum, self._obj)

    def size(self):
        """Number of events or elements in a bin.

        :return: The number of elements in each of the input bins.
        """
        return _call_cpp_func(_cpp.bin_size, self._obj)

    def concatenate(self, other=None, dim=None, out=None):
        """Concatenate bins element-wise by concatenating bin contents along
        their internal bin dimension.

        This can be used as a binary operation, or a reduction operation: The
        bins to concatenate are either obtained element-wise from `self`
        and `other`, or, if `dim` but not `other` is given, from all bins
        along the given dimension.

        :param other: Optional input containing bins.
        :param dim: Optional dimension along which to merge bins. If not given
                    and `other` is `None`, the bins are merged along all
                    dimensions.
        :param out: Optional output buffer.
        :raises: If `other` is not binned data.
        :return: The bins of the two inputs merged.
        """
        if other is not None and dim is not None:
            raise RuntimeError(
                "`concatenate` requires `other` or a `dim`, but not both.")
        if other is not None:
            if out is None:
                return _call_cpp_func(_cpp.buckets.concatenate, self._obj,
                                      other)
            else:
                if self._obj is out:
                    _call_cpp_func(_cpp.buckets.append, self._obj, other)
                else:
                    out = _call_cpp_func(_cpp.buckets.concatenate, self._obj,
                                         other)
                return out
        if out is not None:
            raise RuntimeError(
                "`out` arg not supported for concatenate along dim")
        if dim is not None:
            return _call_cpp_func(_cpp.buckets.concatenate, self._obj, dim)
        raise RuntimeError("Reduction along all dims not supported yet.")


class GroupbyBins:
    """
    Proxy for operations on bins of a groupby object
    """
    def __init__(self, obj):
        self._obj = obj

    def concatenate(self, dim):
        return self._obj.concatenate(dim)


def _bins(obj):
    """
    Returns binning object allowing bin-wise operations
    to be performed.
    """
    if _cpp.is_bins(obj):
        return _Bins(obj)
    else:
        return None


def _set_bins(obj, bins: _Bins):
    # Should only be used by __iadd__ and friends
    assert obj is bins._obj


def _groupby_bins(obj):
    return GroupbyBins(obj)


def histogram(x, bins):
    """Create dense data by histogramming data along all dimension given by
    edges.

    :return: DataArray with values equal to the sum of values in each given
             bin.
    :seealso: :py:func:`scipp.bin` for binning data.
    """
    return _call_cpp_func(_cpp.histogram, x, bins)


def bin(x, edges=None, groups=None, erase=None):
    """Create binned data by binning data along all dimensions given by edges.
    Can specify dimensions with existing binning to erase.

    This does not histogram the data, each output bin will contain a "list" of
    input values.

    :return: Variable containing data in bins.
    :seealso: :py:func:`scipp.histogram` for histogramming data,
              :py:func:`scipp.bins` for creating binned data based on
              explicitly given index ranges.
    """
    if erase is None:
        erase = []
    if groups is None:
        groups = []
    if edges is None:
        edges = []
    return _call_cpp_func(_cpp.bin, x, edges, groups, erase)


def bins(*args, **kwargs):
    """Create binned data, i.e., a variable with elements that are bins.

    The elements of the returned variable are "bins", defined as views into
    `data`. The returned variable keeps and manages a copy of `data`
    internally.

    The variables `begin` and `end` must have the same dims and shape and
    `dtype=sc.dtype.int64`. The output dims and shape are given by `begin`.
    If only `begin` is given, each bucket is a slice containing a non-range
    slice of `data` at the given indices. If neither `begin` nor `end` are
    given, the output has `dims=[dim]` and contains all non-range slices along
    that dimension.

    :param begin: Optional begin indices of bins, used for slicing `data`.
                  If not provided each row of `data` is mapped to a different
                  bin.
    :param end: Optional end indices of bins, used for slicing `data`. If not
                provided this is assumed to be begin + 1.
    :param dim: Dimension of `data` that will be sliced to obtain data for any
                given bin.
    :param data: A variable, data array, or dataset containing combined data
                 of all bins.
    :return: Variable containing data in bins.
    :seealso: :py:func:`scipp.bin` for creating such variables based on
              binning of coord value instead of explicitly given index ranges.
    """
    return _call_cpp_func(_cpp.bins, *args, **kwargs)

File: dataset/bins.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
/// @file
/// @author Simon Heybrock
#include <algorithm>

#include "scipp/common/overloaded.h"
#include "scipp/core/bucket.h"
#include "scipp/core/element/event_operations.h"
#include "scipp/core/element/histogram.h"
#include "scipp/core/except.h"
#include "scipp/core/histogram.h"

#include "scipp/variable/arithmetic.h"
#include "scipp/variable/bins.h"
#include "scipp/variable/bucket_model.h"
#include "scipp/variable/cumulative.h"
#include "scipp/variable/misc_operations.h"
#include "scipp/variable/reduction.h"
#include "scipp/variable/shape.h"
#include "scipp/variable/subspan_view.h"
#include "scipp/variable/transform.h"
#include "scipp/variable/transform_subspan.h"
#include "scipp/variable/variable_factory.h"

#include "scipp/dataset/bin.h"
#include "scipp/dataset/bins.h"
#include "scipp/dataset/bins_view.h"
#include "scipp/dataset/dataset.h"
#include "scipp/dataset/histogram.h"
#include "scipp/dataset/reduction.h"
#include "scipp/dataset/shape.h"

#include "../variable/operations_common.h"
#include "bin_common.h"
#include "dataset_operations_common.h"

namespace scipp::dataset {
namespace {
constexpr auto copy_or_match = [](const auto &a, auto &&b, const Dim dim,
                                  const Variable &srcIndices,
                                  const Variable &dstIndices) {
  if (a.dims().contains(dim))
    copy_slices(a, b, dim, srcIndices, dstIndices);
  else
    core::expect::equals(a, b);
};

constexpr auto expect_matching_keys = [](const auto &a, const auto &b) {
  bool ok = true;
  constexpr auto key = [](const auto &x_) {
    if constexpr (std::is_base_of_v<DataArrayConstView,
                                    std::decay_t<decltype(x_)>>)
      return x_.name();
    else
      return x_.first;
  };
  for (const auto &x : a)
    ok &= b.contains(key(x));
  for (const auto &x : b)
    ok &= a.contains(key(x));
  if (!ok)
    throw std::runtime_error("Mismatching keys in\n" + to_string(a) + " and\n" +
                             to_string(b));
};

} // namespace

void copy_slices(const DataArrayConstView &src, DataArray dst, const Dim dim,
                 const Variable &srcIndices, const Variable &dstIndices) {
  copy_slices(src.data(), dst.data(), dim, srcIndices, dstIndices);
  expect_matching_keys(src.meta(), dst.meta());
  expect_matching_keys(src.masks(), dst.masks());
  for (const auto &[name, coord] : src.meta())
    copy_or_match(coord, dst.meta()[name], dim, srcIndices, dstIndices);
  for (const auto &[name, mask] : src.masks())
    copy_or_match(mask, dst.masks()[name], dim, srcIndices, dstIndices);
}

void copy_slices(const DatasetConstView &src, Dataset dst, const Dim dim,
                 const Variable &srcIndices, const Variable &dstIndices) {
  for (const auto &[name, var] : src.coords())
    copy_or_match(var, dst.coords()[name], dim, srcIndices, dstIndices);
  expect_matching_keys(src.coords(), dst.coords());
  expect_matching_keys(src, dst);
  for (const auto &item : src) {
    const auto &dst_ = dst[item.name()];
    expect_matching_keys(item.attrs(), dst_.attrs());
    expect_matching_keys(item.masks(), dst_.masks());
    copy_or_match(item.data(), dst_.data(), dim, srcIndices, dstIndices);
    for (const auto &[name, var] : item.masks())
      copy_or_match(var, dst_.masks()[name], dim, srcIndices, dstIndices);
    for (const auto &[name, var] : item.attrs())
      copy_or_match(var, dst_.attrs()[name], dim, srcIndices, dstIndices);
  }
}

namespace {
constexpr auto copy_or_resize = [](const auto &var, const Dim dim,
                                   const scipp::index size) {
  auto dims = var.dims();
  if (dims.contains(dim))
    dims.resize(dim, size);
  // Using variableFactory instead of variable::resize for creating
  // _uninitialized_ variable.
  return var.dims().contains(dim)
             ? variable::variableFactory().create(var.dtype(), dims, var.unit(),
                                                  var.hasVariances())
             : copy(var);
};
}

// TODO These functions are an unfortunate near-duplicate of `resize`. However,
// the latter drops coords along the resized dimension. Is there a way to unify
// this? Can the need to drop coords in resize be avoided?
DataArray resize_default_init(const DataArrayConstView &parent, const Dim dim,
                              const scipp::index size) {
  DataArray buffer(copy_or_resize(parent.data(), dim, size));
  for (const auto &[name, var] : parent.coords())
    buffer.coords().set(name, copy_or_resize(var, dim, size));
  for (const auto &[name, var] : parent.masks())
    buffer.masks().set(name, copy_or_resize(var, dim, size));
  for (const auto &[name, var] : parent.attrs())
    buffer.attrs().set(name, copy_or_resize(var, dim, size));
  return buffer;
}

Dataset resize_default_init(const DatasetConstView &parent, const Dim dim,
                            const scipp::index size) {
  Dataset buffer;
  for (const auto &[name, var] : parent.coords())
    buffer.coords().set(name, copy_or_resize(var, dim, size));
  for (const auto &item : parent) {
    buffer.setData(item.name(), copy_or_resize(item.data(), dim, size));
    for (const auto &[name, var] : item.masks())
      buffer[item.name()].masks().set(name, copy_or_resize(var, dim, size));
    for (const auto &[name, var] : item.attrs())
      buffer[item.name()].attrs().set(name, copy_or_resize(var, dim, size));
  }
  return buffer;
}

template <class T>
Variable make_bins_impl(Variable indices, const Dim dim, T &&buffer) {
  indices.setDataHandle(std::make_unique<variable::DataModel<bucket<T>>>(
      indices.data_handle(), dim, std::move(buffer)));
  return indices;
}

/// Construct a bin-variable over a data array.
///
/// Each bin is represented by a Variable slice. `indices` defines the array of
/// bins as slices of `buffer` along `dim`.
Variable make_bins(Variable indices, const Dim dim, DataArray buffer) {
  return make_bins_impl(std::move(indices), dim, std::move(buffer));
}

/// Construct a bin-variable over a dataset.
///
/// Each bin is represented by a Variable slice. `indices` defines the array of
/// bins as slices of `buffer` along `dim`.
Variable make_bins(Variable indices, const Dim dim, Dataset buffer) {
  return make_bins_impl(std::move(indices), dim, std::move(buffer));
}

namespace {
template <class T> Variable bucket_sizes_impl(const Variable &view) {
  const auto &indices = std::get<0>(view.constituents<bucket<T>>());
  const auto [begin, end] = unzip(indices);
  return end - begin;
}
} // namespace

Variable bucket_sizes(const Variable &var) {
  if (var.dtype() == dtype<bucket<Variable>>)
    return bucket_sizes_impl<Variable>(var);
  else if (var.dtype() == dtype<bucket<DataArray>>)
    return bucket_sizes_impl<DataArray>(var);
  else if (var.dtype() == dtype<bucket<Dataset>>)
    return bucket_sizes_impl<Dataset>(var);
  else
    return makeVariable<scipp::index>(var.dims());
}

DataArray bucket_sizes(const DataArrayConstView &array) {
  return {bucket_sizes(array.data()), array.coords(), array.masks(),
          array.attrs()};
}

Dataset bucket_sizes(const DatasetConstView &dataset) {
  return apply_to_items(dataset, [](auto &&_) { return bucket_sizes(_); });
}

bool is_bins(const DataArrayConstView &array) { return is_bins(array.data()); }

bool is_bins(const DatasetConstView &dataset) {
  return std::any_of(dataset.begin(), dataset.end(),
                     [](const auto &item) { return is_bins(item); });
}

} // namespace scipp::dataset

namespace scipp::dataset::buckets {
namespace {

template <class T> auto combine(const Variable &var0, const Variable &var1) {
  const auto &[indices0, dim0, buffer0] = var0.constituents<bucket<T>>();
  const auto &[indices1, dim1, buffer1] = var1.constituents<bucket<T>>();
  static_cast<void>(buffer1);
  static_cast<void>(dim1);
  const Dim dim = dim0;
  const auto [begin0, end0] = unzip(indices0);
  const auto [begin1, end1] = unzip(indices1);
  const auto sizes0 = end0 - begin0;
  const auto sizes1 = end1 - begin1;
  const auto sizes = sizes0 + sizes1;
  const auto end = cumsum(sizes);
  const auto begin = end - sizes;
  const auto total_size =
      end.dims().volume() > 0
          ? end.template values<scipp::index>().as_span().back()
          : 0;
  auto buffer = resize_default_init(buffer0, dim, total_size);
  copy_slices(buffer0, buffer, dim, indices0, zip(begin, end - sizes1));
  copy_slices(buffer1, buffer, dim, indices1, zip(begin + sizes0, end));
  return std::make_shared<variable::DataModel<bucket<T>>>(
      zip(begin, end).data_handle(), dim, std::move(buffer));
}

template <class T>
auto concatenate_impl(const Variable &var0, const Variable &var1) {
  return Variable{merge(var0.dims(), var1.dims()), combine<T>(var0, var1)};
}

template <class T> void reserve_impl(Variable &var, const Variable &shape) {
  // TODO this only reserves in the bins, but assumes buffer has enough space
  auto &&[indices, dim, buffer] = var.constituents<bucket<T>>();
  static_cast<void>(dim);
  static_cast<void>(buffer);
  variable::transform_in_place(
      indices, shape,
      overloaded{
          core::element::arg_list<std::tuple<scipp::index_pair, scipp::index>>,
          core::keep_unit,
          [](auto &begin_end, auto &size) { begin_end.second += size; }});
}

} // namespace

void reserve(Variable &var, const Variable &shape) {
  if (var.dtype() == dtype<bucket<Variable>>)
    return reserve_impl<Variable>(var, shape);
  else if (var.dtype() == dtype<bucket<DataArray>>)
    return reserve_impl<DataArray>(var, shape);
  else
    return reserve_impl<Dataset>(var, shape);
}

Variable concatenate(const Variable &var0, const Variable &var1) {
  if (var0.dtype() == dtype<bucket<Variable>>)
    return concatenate_impl<Variable>(var0, var1);
  else if (var0.dtype() == dtype<bucket<DataArray>>)
    return concatenate_impl<DataArray>(var0, var1);
  else
    return concatenate_impl<Dataset>(var0, var1);
}

DataArray concatenate(const DataArrayConstView &a,
                      const DataArrayConstView &b) {
  return DataArray{
      buckets::concatenate(a.data(), b.data()), union_(a.coords(), b.coords()),
      union_or(a.masks(), b.masks()), intersection(a.attrs(), b.attrs())};
}

/// Reduce a dimension by concatenating all elements along the dimension.
///
/// This is the analogue to summing non-bucket data.
Variable concatenate(const Variable &var, const Dim dim) {
  if (var.dtype() == dtype<bucket<Variable>>)
    return concat_bins<Variable>(var, dim);
  else
    return concat_bins<DataArray>(var, dim);
}

/// Reduce a dimension by concatenating all elements along the dimension.
///
/// This is the analogue to summing non-bucket data.
DataArray concatenate(const DataArrayConstView &array, const Dim dim) {
  return groupby_concat_bins(array, {}, {}, {dim});
}

void append(Variable &var0, const Variable &var1) {
  if (var0.dtype() == dtype<bucket<Variable>>)
    var0.setDataHandle(combine<Variable>(var0, var1));
  else if (var0.dtype() == dtype<bucket<DataArray>>)
    var0.setDataHandle(combine<DataArray>(var0, var1));
  else
    var0.setDataHandle(combine<Dataset>(var0, var1));
}

void append(Variable &&var0, const Variable &var1) { append(var0, var1); }

void append(DataArray &a, const DataArrayConstView &b) {
  expect::coordsAreSuperset(a, b);
  union_or_in_place(a.masks(), b.masks());
  append(a.data(), b.data());
}

Variable histogram(const Variable &data, const Variable &binEdges) {
  using namespace scipp::core;
  auto hist_dim = binEdges.dims().inner();
  auto &&[indices, dim, buffer] = data.constituents<bucket<DataArray>>();
  // `hist_dim` may be the same as a dim of data if there is existing binning.
  // We rename to a dummy to avoid duplicate dimensions, perform histogramming,
  // and then sum over the dummy dimensions, i.e., sum contributions from all
  // inputs bins to the same output histogram. This also allows for threading of
  // 1-D histogramming provided that the input has multiple bins along
  // `hist_dim`.
  std::string nonclashing_name("dummy");
  for (const auto &d : indices.dims().labels())
    nonclashing_name += d.name();
  const Dim dummy = Dim(nonclashing_name);
  indices.rename(hist_dim, dummy);
  const Masker masker(buffer, dim);
  auto hist = variable::transform_subspan(
      buffer.dtype(), hist_dim, binEdges.dims()[hist_dim] - 1,
      subspan_view(buffer.meta()[hist_dim], dim, indices),
      subspan_view(masker.data(), dim, indices), binEdges, element::histogram);
  if (hist.dims().contains(dummy))
    return sum(hist, dummy);
  else
    return hist;
}

Variable map(const DataArrayConstView &function, const Variable &x, Dim dim) {
  if (dim == Dim::Invalid)
    dim = edge_dimension(function);
  const Masker masker(function, dim);
  const auto &coord = bins_view<DataArray>(x).meta()[dim];
  const auto &edges = function.meta()[dim];
  const auto weights = subspan_view(masker.data(), dim);
  if (all(islinspace(edges, dim)).value<bool>()) {
    return variable::transform(coord, subspan_view(edges, dim), weights,
                               core::element::event::map_linspace);
  } else {
    if (!issorted(edges, dim))
      throw except::BinEdgeError("Bin edges of histogram must be sorted.");
    return variable::transform(coord, subspan_view(edges, dim), weights,
                               core::element::event::map_sorted_edges);
  }
}

void scale(DataArray &array, const DataArrayConstView &histogram, Dim dim) {
  if (dim == Dim::Invalid)
    dim = edge_dimension(histogram);
  // Coords along dim are ignored since "binning" is dynamic for buckets.
  expect::coordsAreSuperset(array, histogram.slice({dim, 0}));
  // scale applies masks along dim but others are kept
  union_or_in_place(array.masks(), histogram.slice({dim, 0}).masks());
  const Masker masker(histogram, dim);
  auto data = bins_view<DataArray>(array.data()).data();
  const auto &coord = bins_view<DataArray>(array.data()).meta()[dim];
  const auto &edges = histogram.meta()[dim];
  const auto weights = subspan_view(masker.data(), dim);
  if (all(islinspace(edges, dim)).value<bool>()) {
    transform_in_place(data, coord, subspan_view(edges, dim), weights,
                       core::element::event::map_and_mul_linspace);
  } else {
    if (!issorted(edges, dim))
      throw except::BinEdgeError("Bin edges of histogram must be sorted.");
    transform_in_place(data, coord, subspan_view(edges, dim), weights,
                       core::element::event::map_and_mul_sorted_edges);
  }
}

namespace {
Variable applyMask(const DataArrayConstView &buffer, const Variable &indices,
                   const Dim dim, const Variable &masks) {
  auto indices_copy = Variable(indices);
  auto masked_data = scipp::variable::masked_to_zero(buffer.data(), masks);
  return make_bins(std::move(indices_copy), dim, std::move(masked_data));
}

} // namespace

Variable sum(const Variable &data) {
  auto type = variable::variableFactory().elem_dtype(data);
  type = type == dtype<bool> ? dtype<int64_t> : type;
  const auto unit = variable::variableFactory().elem_unit(data);
  Variable summed;
  if (variable::variableFactory().hasVariances(data))
    summed = Variable(type, data.dims(), unit, Values{}, Variances{});
  else
    summed = Variable(type, data.dims(), unit, Values{});

  if (data.dtype() == dtype<bucket<DataArray>>) {
    const auto &&[indices, dim, buffer] =
        data.constituents<bucket<DataArray>>();
    if (const auto mask_union = irreducible_mask(buffer.masks(), dim)) {
      variable::sum_impl(summed, applyMask(buffer, indices, dim, mask_union));
    } else {
      variable::sum_impl(summed, data);
    }
  } else {
    variable::sum_impl(summed, data);
  }

  return summed;
}

DataArray sum(const DataArrayConstView &data) {
  return {buckets::sum(data.data()), data.coords(), data.masks(), data.attrs()};
}

Dataset sum(const DatasetConstView &d) {
  return apply_to_items(d, [](auto &&... _) { return buckets::sum(_...); });
}

} // namespace scipp::dataset::buckets
